src/ucm/util/reloc.h

```diff
@@ -46,4 +46,72 @@ ucs_status_t ucm_reloc_modify(ucm_reloc_patch_t* patch);
 void* ucm_reloc_get_orig(const char *symbol, void *replacement);
 
 
+/**
+ * Define a replacement function to a memory-mapping function call, which calls
+ * the event handler, and if event handler returns error code - calls the original
+ * function.
+ */
+#define UCM_DEFINE_REPLACE_FUNC(_category, _name, _rettype, _fail_val, ...) \
+    \
+    _rettype ucm_override_##_name(UCM_FUNC_DEFINE_ARGS(__VA_ARGS__)); \
+    \
+    /* Call the original function using dlsym(RTLD_NEXT) */ \
+    _rettype ucm_orig_##_name(UCM_FUNC_DEFINE_ARGS(__VA_ARGS__)) \
+    { \
+        typedef _rettype (*func_ptr_t) (__VA_ARGS__); \
+        static func_ptr_t orig_func_ptr = NULL; \
+        \
+        ucm_trace("%s()", __FUNCTION__); \
+        \
+        if (ucs_unlikely(orig_func_ptr == NULL)) { \
+            pthread_mutex_lock(&ucm_##_category##_get_orig_lock); \
```

Comment:
do we really need a separate lock for cuda?

Reply:
not really. but not sure if I can unify lock for both mmap and cuda

Reply:
i think that if the lock will not be split, there would not be a need to define separate macros for cuda and mmap

Reply:
unified lock and used same macro for cuda and mmap